lessons/230/elixir-app/lib/app/router.ex

```diff
@@ -51,44 +45,46 @@ defmodule App.Router do
       }
     ]
 
-    conn
-    |> put_resp_content_type("application/json")
-    |> send_resp(200, Jason.encode!(devices))
-    |> halt()
+    json(conn, 200, devices)
   end
 
   post "/api/devices" do
     try do
       {:ok, body, conn} = Plug.Conn.read_body(conn)
       body = Jason.decode!(body)
-      now = DateTime.utc_now()
+      now = DateTime.utc_now() |> DateTime.to_string()
 
       device =
-        App.Repo.insert!(%App.Device{
+        App.Device.save(%App.Device{
           uuid: Ecto.UUID.generate(),
           mac: body["mac"],
           firmware: body["firmware"],
           created_at: now,
           updated_at: now
         })
 
-      conn
-      |> put_resp_content_type("application/json")
-      |> send_resp(201, Jason.encode!(device))
-      |> halt()
+      json(conn, 201, device)
     rescue
       e ->
-        conn
-        |> put_resp_content_type("application/json")
-        |> send_resp(400, Jason.encode!(%{message: Exception.message(e)}))
-        |> halt()
+        json(conn, 400, %{message: Exception.message(e)})
     end
   end
 
   match _ do
+    text(conn, 404, "Not found")
+  end
+
+  defp text(conn, status, body) do
     conn
     |> put_resp_content_type("text/plain")
-    |> send_resp(404, "Not found")
+    |> send_resp(status, body)
+    |> halt()
+  end
+
+  defp json(conn, status, payload) do
+    conn
+    |> put_resp_content_type("application/json")
+    |> send_resp(status, Jason.encode_to_iodata!(payload))
```

Comment:
Refactor and use the more efficient `Jason.encode_to_iodata!`.

Reply:
There are the types of optimizations I avoided since I cared more about the "common" situations. Primarily because they aren't the default ones 😭 

I am glad you are aware of the test; now we get to see the impact of this extra stuff.

💜 🚀

Reply:
What do you mean this isn't the default one? That's precisely how Phoenix would encode it and it is the recommended way. I don't think there is any value in doing something different than what the most common web framework would do.

Reply:
FWIW I've also seen quite often that `Jason.encode!` was used instead of `Jason.encode_to_iodata!` in "user-space" code. I think the misconception comes from the name, people see `encode` and don't look any further ...

- https://github.com/search?q=Jason.encode_to_iodata%21+language%3AElixir&type=code
- https://github.com/search?q=Jason.encode%21+language%3AElixir&type=code

And I'm glad `:json` went with `encode` returning iodata by default :)

Reply:
@josevalim It is not that important, I was curious about baseline of what I seem based on my career, before adding every optimization in the book. The obvious one is to remove Ecto, as an example.

I am also curious to add cowboy to the mix 🤔 just out of curiosity

Reply:
@yordis the issue is that, by not making them as close as possible, wrong conclusions may be derived from the data. Ecto should have not been used for inserts from the beginning IMO. :) Regarding Cowboy, Bandit has been in my experience consistently faster.

@ruslandoga unfortunately Elixir's new JSON mirrors Jason. The trade-offs here are tricky, because if `JSON.encode!` returned iodata, people would be very confused and it may even be the incorrect result in some cases. I have added a clear warning/disclaimer to the docs though. :)

Reply:
The intent was to have multiple runs, I am interested in seeing the differences load test, that is all

Reply:
@yordis @josevalim - I'd also like to see Cowboy as well. The reason I mention this is because in my own testing when hitting both servers with load testing scripts, I saw much higher CPU usage from Bandit than Cowboy. Like much much higher at the limit. This was about 9 months ago, I might see if I can reproduce this and if so might ask in the forum for some help trying to discover the reason. ....assuming I can reproduce it.

Reply:
This is a rabbit hole that we love 😄 we will do all possible combinations to find the fastest one!

Reply:
Bandit is generally faster than Cowboy, so I am not expecting large changes here.